src/Common/src/TypeSystem/Common/TypeSystemException.cs

```diff
@@ -0,0 +1,64 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+
+namespace Internal.TypeSystem
+{
+    /// <summary>
+    /// Base type for all type system exceptions.
+    /// </summary>
+    public abstract class TypeSystemException : Exception
+    {
+        public TypeSystemException(string message)
+            : base(message)
+        {
+        }
+
+        /// <summary>
+        /// The exception that is thrown when type-loading failures occur.
+        /// </summary>
+        public class TypeLoadException : TypeSystemException
+        {
+            internal TypeLoadException(string message)
+                : base(message)
+            {
+            }
+        }
+
+        /// <summary>
+        /// The exception that is thrown when there is an attempt to access a class member that does not exist
+        /// or that is not declared as public.
+        /// </summary>
+        public abstract class MissingMemberException : TypeSystemException
+        {
+            protected internal MissingMemberException(string message)
+                : base(message)
```

Comment:
> This sounds like days of work fighting buildtools and it will likely become a nightmare for loc tooling too

The loc thing can ne done later. `switch` over ID to return the hardcoded message would be fine as interim solution. 

It may be useful to work towards the message ID + parameters, e.g. make them fields of the TypeSystemException - if you prefer to not do this in this change, it is fine with me too.